sdks/python/apache_beam/pipeline.py

```diff
@@ -191,47 +189,73 @@ class TransformUpdater(PipelineVisitor): # pylint: disable=used-before-assignmen
-      def _replace_if_needed(self, transform_node):
-        if matcher(transform_node):
+      def _replace_if_needed(self, original_transform_node):
+        if matcher(original_transform_node):
+          assert isinstance(original_transform_node, AppliedPTransform)
           replacement_transform = override.get_replacement_transform(
-              transform_node.transform)
-          inputs = transform_node.inputs
-          # TODO:  Support replacing PTransforms with multiple inputs.
+              original_transform_node.transform)
+
+          replacement_trasform_node = AppliedPTransform(
+              original_transform_node.parent, replacement_transform,
+              original_transform_node.full_label,
+              original_transform_node.inputs)
+
+          # Transform execution could depend on order in which nodes are
+          # considered. Hence we insert the replacement transform node to same
+          # index as the original transform node. Note that this operation
+          # removes the original transform node.
+          if original_transform_node.parent:
+            assert isinstance(original_transform_node.parent, AppliedPTransform)
+            parent_parts = original_transform_node.parent.parts
+            parent_parts[parent_parts.index(original_transform_node)] = (
+                replacement_trasform_node)
+          else:
+            # Original transform has to be a root
+            roots = self.pipeline.transforms_stack[0].parts
+            assert original_transform_node in roots
+            roots[roots.index(original_transform_node)] = (
+                replacement_trasform_node)
+
+          inputs = replacement_trasform_node.inputs
+          # # TODO:  Support replacing PTransforms with multiple inputs.
           if len(inputs) > 1:
             raise NotImplementedError(
                 'PTransform overriding is only supported for PTransforms that '
                 'have a single input. Tried to replace input of '
                 'AppliedPTransform %r that has %d inputs',
-                transform_node, len(inputs))
-          transform_node.transform = replacement_transform
-          self.pipeline.transforms_stack.append(transform_node)
+                original_transform_node, len(inputs))
+
+          # We have to add the new AppliedTransform to the stack before expand()
+          # and pop it out later to make sure that parts get added correctly.
+          self.pipeline.transforms_stack.append(replacement_trasform_node)
 
           # Keeping the same label for the replaced node but recursively
-          # removing labels of child transforms since they will be replaced
-          # during the expand below.
-          self.pipeline._remove_labels_recursively(transform_node)
+          # removing labels of child transforms of original transform since they
+          # will be replaced during the expand below.
+          # This is needed in-case replacement contains children that have
```

Comment:
<!--thread_id:cc_164903019_t; commit:eb3d1308494b34cb40336a41f40937e9b74c266a; resolved:1-->
<!--section:context-quote-->
> **charlesccychen** wrote:
> Add newline (or merge with previous line).
> 
> "in case the replacement"

<!--section:body-->
Done.